torch/csrc/jit/tensorexpr/llvm_codegen.h

```diff
@@ -3,117 +3,32 @@
 #ifdef ENABLE_LLVM
 #include <torch/csrc/WindowsTorchApiMacro.h>
 
-#include "llvm/ExecutionEngine/Orc/ThreadSafeModule.h"
 #include "torch/csrc/jit/tensorexpr/codegen.h"
 #include "torch/csrc/jit/tensorexpr/ir.h"
 #include "torch/csrc/jit/tensorexpr/ir_visitor.h"
-#include "torch/csrc/jit/tensorexpr/llvm_jit.h"
 
-#include <llvm/IR/IRBuilder.h>
-#include <llvm/IR/LegacyPassManager.h>
-#include <llvm/IR/Verifier.h>
 #include <unordered_map>
 #include <vector>
 
-#define DEBUG_PRINT 0
-
-#if DEBUG_PRINT
-#include <llvm/IR/LegacyPassManager.h>
-#endif
-
 namespace torch {
 namespace jit {
 namespace tensorexpr {
 
-class TORCH_API LLVMCodeGen : public CodeGen, public IRVisitor {
- private:
-  llvm::orc::ThreadSafeContext context_;
-  llvm::IRBuilder<> irb_;
-  std::unique_ptr<llvm::TargetMachine> TM_;
-  std::unique_ptr<llvm::orc::PytorchLLVMJIT> jit_;
-  std::unique_ptr<llvm::Module> module_;
-  llvm::Function* fn_;
-  llvm::BasicBlock* bb_;
-  llvm::Value* value_;
-  llvm::JITTargetAddress kernelAddress_;
-
-#define LLVM_TYPE_DECLARE(_1, Name) llvm::Type* Name##Ty_;
-  AT_FORALL_SCALAR_TYPES_AND2(Bool, Half, LLVM_TYPE_DECLARE);
-#undef LLVM_TYPE_DECLARE
-
-  std::unordered_map<const Var*, int> varToArg_;
-  std::unordered_map<const Var*, llvm::Value*> varToVal_;
-
-  std::vector<void*> args_;
-
- private:
-  llvm::LLVMContext& getContext();
-  llvm::Type* dtypeToLLVM(Dtype dtype);
-  llvm::Type* dtypeToLLVMPtr(Dtype dtype);
-  void emitWrapper(const std::vector<llvm::Type*>& params);
-  void emitKernel(Stmt* stmt, const std::vector<llvm::Type*>& params);
+class LLVMCodeGenImpl;
 
+class TORCH_API LLVMCodeGen : public CodeGen {
  public:
   explicit LLVMCodeGen(
       Stmt* stmt,
       const std::vector<BufferArg>& args,
       Dtype dtype = kInt);
   explicit LLVMCodeGen(Stmt* stmt);
 
-  ~LLVMCodeGen() override {}
+  LLVMCodeGen() = delete;
```

Comment:
minor nit: I think this is automatically deleted, with the first constructor. What is not disabled is the copy and move constructors.

Reply:
Yeah it's possible the only change i actually needed here was moving the dtor to the .cpp.  I'll try that out.